packages/backend/src/tokens/request.ts

```diff
@@ -688,22 +796,148 @@ ${error.getFullMessage()}`,
+    const mismatchState = maybeHandleTokenTypeMismatch(parsedTokenType, acceptsToken, authenticateContext);
+    if (mismatchState) {
+      return mismatchState;
+    }
+
+    const { data, tokenType, errors } = await verifyMachineAuthToken(
+      sessionOrMachineTokenInHeader,
+      authenticateContext,
+    );
+    if (errors) {
+      return handleMachineError(tokenType, errors[0]);
+    }
+    return signedIn({
+      tokenType,
+      authenticateContext,
+      machineData: data,
+      token: sessionOrMachineTokenInHeader,
+    });
+  }
+
+  async function authenticateAnyRequestWithTokenInHeader() {
+    const { sessionOrMachineTokenInHeader } = authenticateContext;
+    // Use session token error handling if no token in header (default behavior)
+    if (!sessionOrMachineTokenInHeader) {
+      return handleError(new Error('Missing token in header'), 'header');
+    }
+
+    // Handle as a machine token
+    if (isMachineToken(sessionOrMachineTokenInHeader)) {
+      const parsedTokenType = getMachineTokenType(sessionOrMachineTokenInHeader);
+      const mismatchState = maybeHandleTokenTypeMismatch(parsedTokenType, acceptsToken, authenticateContext);
+      if (mismatchState) {
+        return mismatchState;
+      }
+
+      const { data, tokenType, errors } = await verifyMachineAuthToken(
+        sessionOrMachineTokenInHeader,
+        authenticateContext,
+      );
+      if (errors) {
+        return handleMachineError(tokenType, errors[0]);
+      }
+
+      return signedIn({
+        tokenType,
+        authenticateContext,
+        machineData: data,
+        token: sessionOrMachineTokenInHeader,
+      });
+    }
+
+    // Handle as a regular session token
+    const { data, errors } = await verifyToken(sessionOrMachineTokenInHeader, authenticateContext);
+    if (errors) {
+      return handleError(errors[0], 'header');
+    }
+
+    return signedIn({
+      tokenType: TokenType.SessionToken,
+      authenticateContext,
+      sessionClaims: data,
+      token: sessionOrMachineTokenInHeader,
+    });
   }
 
-  if (authenticateContext.sessionTokenInHeader) {
-    return authenticateRequestWithTokenInHeader();
+  if (authenticateContext.sessionOrMachineTokenInHeader) {
+    if (acceptsToken === 'any') {
+      return authenticateAnyRequestWithTokenInHeader();
+    }
+
+    if (acceptsToken === TokenType.SessionToken) {
+      return authenticateRequestWithTokenInHeader();
+    }
+
+    return authenticateMachineRequestWithTokenInHeader();
+  }
+
+  // Machine requests cannot have the token in the cookie, it must be in header.
+  if (
+    acceptsToken === TokenType.OAuthToken ||
+    acceptsToken === TokenType.ApiKey ||
+    acceptsToken === TokenType.MachineToken
+  ) {
+    return signedOut({
+      tokenType: acceptsToken,
+      authenticateContext,
+      reason: 'No token in header',
+    });
   }
 
   return authenticateRequestWithTokenInCookie();
-}
+}) as AuthenticateRequest;
```

Comment:
without this assertion, TS would error because it can't prove our single implementation correctly handles all the specific type cases defined in the interface overloads